src/execution/execute.ts

```diff
@@ -790,106 +791,128 @@ function completeLeafValue(
  */
 function completeAbstractValue(
   exeContext: ExecutionContext,
-  returnType: GraphQLAbstractType,
+  abstractType: GraphQLAbstractType,
   fieldNodes: ReadonlyArray<FieldNode>,
   info: GraphQLResolveInfo,
   path: Path,
   result: unknown,
 ): PromiseOrValue<ObjMap<unknown>> {
-  const resolveTypeFn = returnType.resolveType ?? exeContext.typeResolver;
+  const resolveTypeFn = abstractType.resolveType ?? exeContext.typeResolver;
   const contextValue = exeContext.contextValue;
-  const runtimeType = resolveTypeFn(result, contextValue, info, returnType);
-
-  if (isPromise(runtimeType)) {
-    return runtimeType.then((resolvedRuntimeType) =>
-      completeObjectValue(
-        exeContext,
-        ensureValidRuntimeType(
-          resolvedRuntimeType,
-          exeContext,
-          returnType,
-          fieldNodes,
-          info,
-          result,
-        ),
-        fieldNodes,
-        info,
-        path,
-        result,
-      ),
-    );
-  }
+  const possibleRuntimeTypeName = resolveTypeFn(
+    result,
+    contextValue,
+    info,
+    abstractType,
+  );
 
-  return completeObjectValue(
+  return completeAbstractValueImpl(
     exeContext,
-    ensureValidRuntimeType(
-      runtimeType,
-      exeContext,
-      returnType,
-      fieldNodes,
-      info,
-      result,
-    ),
+    abstractType,
+    possibleRuntimeTypeName,
     fieldNodes,
     info,
     path,
     result,
   );
 }
 
-function ensureValidRuntimeType(
-  runtimeTypeName: unknown,
+function completeAbstractValueImpl(
   exeContext: ExecutionContext,
-  returnType: GraphQLAbstractType,
+  abstractType: GraphQLAbstractType,
+  possibleRuntimeTypeName: PromiseOrValue<string | undefined>,
   fieldNodes: ReadonlyArray<FieldNode>,
   info: GraphQLResolveInfo,
+  path: Path,
   result: unknown,
-): GraphQLObjectType {
-  if (runtimeTypeName == null) {
+): PromiseOrValue<ObjMap<unknown>> {
+  if (isPromise(possibleRuntimeTypeName)) {
+    return possibleRuntimeTypeName.then((resolved) =>
+      completeAbstractValueImpl(
+        exeContext,
+        abstractType,
+        resolved,
+        fieldNodes,
+        info,
+        path,
+        result,
+      ),
+    );
+  }
+
+  if (possibleRuntimeTypeName == null) {
     throw new GraphQLError(
-      `Abstract type "${returnType.name}" must resolve to an Object type at runtime for field "${info.parentType.name}.${info.fieldName}". Either the "${returnType.name}" type should provide a "resolveType" function or each possible type should provide an "isTypeOf" function.`,
+      `Abstract type "${abstractType.name}" must resolve to an Object type or an intermediate Interface type at runtime for field "${info.parentType.name}.${info.fieldName}". Either the "${abstractType.name}" type should provide a "resolveType" function or each possible type should provide an "isTypeOf" function.`,
```

Comment:
What do you think about only altering this error message in the case that an intermediate Interface type actually exists as a potential case?

Looking at the unit test impact, my sense is the primary downside of this change is the additional complication to this error message that makes it slightly harder to understand.

Ideally for the large majority common case where there is no intermediate interface, we don't burden the error message

Reply:
👍  Tailoring the error message to the interface at hand seems pretty reasonable.